src/web100srv.c

```diff
@@ -1951,375 +1837,335 @@ void NDT_server_main_loop(SSL_CTX *ssl_context, int listenfd) {
+        for (i = 0; i < 5; i++) {
+          child_signal = SRV_QUEUE_TEST_STARTS_NOW;
+          retcode = write(mchild->pipe, &child_signal, sizeof(int));
+          log_println(6, "write(%d) returned %d, errno=%d", mchild->pid,
+                      retcode, errno);
+          if (retcode == sizeof(int)) break;
+          if ((retcode == -1) && (errno == EINTR)) continue;
+          log_println(5,
+                      "Failed to write 'GO' message to client %d, "
+                      "reason=%d, errno=%d",
+                      mchild->pid, retcode, errno);
+          // TODO: handle other error conditions
+          if (retcode == -1) {
+            log_println(1, "Dispatch multi-client failed because '%s'",
+                        strerror(errno));
+            close(mchild->pipe);
+            kill(chld_pid, SIGTERM);
+            goto mainloop;
           }
-          close(mchild->pipe);
-          close(mchild->ctlsockfd);
-        } else {  // multiple client loop complete
-          head_ptr->stime = time(0);
-          head_ptr->running = 1;
-          snprintf(tmpstr, sizeof(tmpstr), "go %d %s", t_opts,
-                   head_ptr->tests);
-          log_println(5, "sending 'GO' signal to client msg='%s'",
-                      tmpstr);
-          send_json_message(head_ptr->ctlsockfd, SRV_QUEUE, "0", 1,
-                            testopt.connection_flags, JSON_SINGLE_VALUE);
-          for (i = 0; i < 5; i++) {
-            retcode = write(head_ptr->pipe, tmpstr, strlen(tmpstr));
-            if ((retcode == -1) && (errno == EINTR))
-              continue;
-            if (retcode == strlen(tmpstr))
-              break;
-            /* TODO: handle other error conditions */
-            if (retcode == -1) {
-              log_println(1, "Dispatch multi-client failed because '%s'",
-                          strerror(errno));
-              shutdown(head_ptr->ctlsockfd, SHUT_WR);
-              close(head_ptr->ctlsockfd);
-              kill(chld_pid, SIGTERM);
-              goto mainloop;
-            }
+        }
+        close(mchild->pipe);
+      } else {  // multiple client loop complete
+        head_ptr->stime = time(0);
+        head_ptr->running = 1;
+        log_println(5, "sending 'GO' signal to client");
+        for (i = 0; i < 5; i++) {
+          child_signal = SRV_QUEUE_TEST_STARTS_NOW;
+          retcode = write(mchild->pipe, &child_signal, sizeof(int));
+          if (retcode == sizeof(int)) break;
+          if ((retcode == -1) && (errno == EINTR)) continue;
+          // TODO: handle other error conditions
+          if (retcode == -1) {
+            log_println(1, "Dispatch multi-client failed because '%s'",
+                        strerror(errno));
+            close(head_ptr->pipe);
+            kill(chld_pid, SIGTERM);
+            goto mainloop;
           }
-          close(head_ptr->pipe);
-          close(head_ptr->ctlsockfd);
         }
```

Comment:
ctlsockfd is no longer used by the parent.  That's actually the main reason I had to refactor everything.